lampshade.go

```diff
@@ -295,13 +295,16 @@ type StatsTracking interface {
 	EMARTT() time.Duration
 }
 
+// DialFN is a function that dials the server
+type DialFN func() (net.Conn, error)
+
 // Dialer is a dialer
 type Dialer interface {
 	StatsTracking
 
-	Dial() (net.Conn, error)
+	Dial(dial DialFN) (net.Conn, error)
```

Comment:
It is a separate op, but it happens in the context of the other op. i.e. dialing a physical connection is a side-effect of requesting a virtual connection when there is no physical connection yet.